psi4/src/psi4/libfock/DirectDFJK.cc

```diff
@@ -0,0 +1,1600 @@
+	//  inverse of the coulomb metric and the (-0.5) power of the metric, so
+	//  we will prepare those immediately afterwards
+	//prepare_metric();
+	get_met();
+	
+	prepare_metric_power(-0.5);
+
+	//for what it's worth, we should have the density matrix
+	//prepare_D_symm();
+
+}
+
+void DirectDFJK::sparsity_prep_pQq(){
+	int procs = 1;
+#ifdef _OPENMP
+	procs = omp_nthread_;
+#endif
+
+	// variables to hold data
+	double global_max_int = 0.0;
+
+
+	std::unique_ptr<double[]> shel_maxes(new double[p_shells_*p_shells_]);
+	std::unique_ptr<double[]> integrals(new double[nbf_*nbf_]);
+
+
+	for (size_t shel_sq_iter = 0; shel_sq_iter < p_shells_*p_shells_; shel_sq_iter++) shel_maxes[shel_sq_iter] = 0.0;
+
+
+	for (size_t func_sq_iter = 0; func_sq_iter < nbf_*nbf_; func_sq_iter++) integrals[func_sq_iter] = 0.0;
+
+	// AO machinery
+	auto rifactory = std::make_shared<IntegralFactory>(primary_, primary_, primary_, primary_);
+
+	std::vector<std::shared_ptr<TwoBodyAOInt>> eri(procs);
+	std::vector<const double*> buffer(procs);
+
+
+#pragma omp parallel num_threads(procs)
+    {
+    int rank = 0;
+#ifdef _OPENMP
+    rank = omp_get_thread_num();
+#endif
+    eri[rank] = std::shared_ptr<TwoBodyAOInt>(rifactory->eri());
+	buffer[rank] = eri[rank]->buffer();
+    }
+
+	// Sparsity member data
+	schwarz_func_ints_.resize(nbf_);
+
+	std::vector<size_t> dense_its(nbf_);
+
+	//AO shell metadata
+	size_t Shell_M_count;
+	size_t Shell_M_start;
+	size_t Shell_N_count;
+	size_t Shell_N_start;
+	double val = 0.0;
+
+#pragma omp parallel for schedule(guided) num_threads(procs) private( Shell_M_count, Shell_M_start, Shell_N_count, Shell_N_start, val) reduction(max:global_max_int)
+	for(size_t Shell_M = 0; Shell_M < p_shells_; Shell_M++) {
+		int rank = 0;
+#ifdef _OPENMP
+    	rank = omp_get_thread_num();
+#endif
+        Shell_M_start = primary_->shell(Shell_M).function_index();
+		Shell_M_count = primary_->shell(Shell_M).nfunction();
+		for (size_t Shell_N = 0; Shell_N < p_shells_; Shell_N++) {
+	        Shell_N_start = primary_->shell(Shell_N).function_index();
+			Shell_N_count = primary_->shell(Shell_N).nfunction();
+			eri[rank]->compute_shell(Shell_M, Shell_N, Shell_M, Shell_N);
+			for (size_t func_m = 0; func_m < Shell_M_count; func_m++){
+				for (size_t func_n = 0; func_n < Shell_N_count; func_n++){
+					integrals[ (Shell_M_start + func_m) * nbf_ + Shell_N_start + func_n ]
+					=
+//					integrals[ (Shell_N_start + func_n) * nbf_ + Shell_M_start + func_m ]
+//					=
+					val 
+					= 
+ fabs(buffer[rank][ func_m * Shell_N_count * Shell_M_count * Shell_N_count + 
+				    func_n * Shell_M_count * Shell_N_count + 
+				    func_m * Shell_N_count + 
+				    func_n]);
+					if (global_max_int < val) global_max_int = val;
+					
+					if (shel_maxes[Shell_M * p_shells_ + Shell_N] < val) {
+						shel_maxes[Shell_M * p_shells_ + Shell_N]
+//						=
+//						shel_maxes[Shell_N * p_shells_ + Shell_M]
+						= val;
+					}
+				}
+			}
+		}
+	}
```

Comment:
So, `integrals` is just the Schwarz screening matrix? You should use a more descriptive name.

Reply:
ok